### ISSUE TITLE
If client secret contains character '+', it will cause invalid client error.

### ISSUE THREAD

**Issue by supermanHub** [panva/node-oidc-provider#361]
**Describe the bug**
If there is symbol `+` in client secret, it will cause `client authentication failed`, the detail error information is 'invalid secret provided'. 

**To Reproduce**
Provider and Client configuration: (inline or gist)
```js
// provider configuration
// normal provider configuration

// keystore(s) - if they affect the behaviour
// normal provider configuration

// client(s) configuration
{
  client_id: a4f13611-c393-4a21-adde-f8c2176c30c2,
  client_secret: 5Q7UtUI6IMFsxLr9J+xicImeXVbuXQFIssDjJIf2oUNQrPCyL1K0KlLb3c1GjWnC
 .
 .
 .
}
```

Steps to reproduce the behaviour:

1.  Dynamic register a client, and make sure there is character `+` in the `client_secret` field of registration response. Or you can use a inmemory client with `client_id` is `foo`, `client_secret` is `foo+bar`.
2.  Send authentication request to `RP`(Login) with upon `client_id`.
3.  An `invalid_client` error will be occurred.

**Expected behaviour**
`OP` should be able to  authenticate the client, even there is a character `+` in client's `client_secret` field.

**Environment:**
 - oidc-provider version: [4.8.1]
 - node version: [8.12.0 LTS]

**Additional context**
The bug is caused by replace `+` with space` ` when decoding the `client_secret`.
````
// In lib/shared/token_auth.js, oidc provider decode the client_secret in this way:
// replace + with space
function decodeAuthToken(token) {
  return decodeURIComponent(token.replace(/\+/g, '%20'));
}
````
and decode result is `5Q7UtUI6IMFsxLr9J xicImeXVbuXQFIssDjJIf2oUNQrPCyL1K0KlLb3c1GjWnC` or `foo bar`.
But in token_credential_auth.js, it compare the `client_secret` in this way
````
// expected is: 5Q7UtUI6IMFsxLr9J+xicImeXVbuXQFIssDjJIf2oUNQrPCyL1K0KlLb3c1GjWnC
// actual is: 5Q7UtUI6IMFsxLr9J xicImeXVbuXQFIssDjJIf2oUNQrPCyL1K0KlLb3c1GjWnC
// They are not match, and then throw InvalidClientAuth error
if (!constantEquals(expected, actual, 1000)) {
    throw new InvalidClientAuth('invalid secret provided');
}
````

**Logs**
`uuid=ea645451-1072-4dfd-af70-6eab2a033e62 path=/token method=POST error={ error: 'invalid_client', error_description: 'client authentication failed' } detail=invalid secret provided`

**Comment by supermanHub** [comment]
There is a high probability of existing character '+' in the client_secret field, if you dynamic register a client with default `defaultSecretFactory (in class: lib/actions/registration.sj)`. 

I don't understand why we have to replace `+` with space ' ' when decoding the `client_secret`. Should it be `%2B`, not `%20`?

**Comment by panva** [comment]
This is actually conform behaviour and not a bug. It was fixed from what you're expecting in 3.0.0. The client_secret_basic is not 100% basic http auth, the username and password tokens are supposed to be additionally formencoded.

A proper way of submitting `client_id` and `client_secret` using `client_secret_basic` is `Authorization: base64(formEncode(client_id):formEncode(client_secret))` as per https://tools.ietf.org/html/rfc6749#section-2.3.1 incl. https://tools.ietf.org/html/rfc6749#appendix-B
